File: services/academic/src/main/java/org/endipi/academic/service/PrerequisiteCourseService.java
package org.endipi.academic.service;

import org.endipi.academic.dto.request.PrerequisiteCourseRequest;
import org.endipi.academic.dto.request.UpdatePrerequisitesRequest;
import org.endipi.academic.dto.response.CoursePrerequisiteDetailsResponse;
import org.endipi.academic.dto.response.CoursePrerequisiteSummaryResponse;
import org.endipi.academic.dto.response.PrerequisiteCourseResponse;

import java.util.List;

public interface PrerequisiteCourseService {
    List<PrerequisiteCourseResponse> findAll();

    PrerequisiteCourseResponse findById(Long id);

    PrerequisiteCourseResponse saveWithRetry(PrerequisiteCourseRequest prerequisiteCourseRequest);

    void deleteById(Long id);
    
    // Enhanced methods for prerequisite management
    List<CoursePrerequisiteSummaryResponse> getAllCoursesWithPrerequisiteInfo();
    
    CoursePrerequisiteDetailsResponse getCoursePrerequisiteDetails(Long courseId);
    
    void updateCoursePrerequisites(UpdatePrerequisitesRequest request);
    
    List<CoursePrerequisiteDetailsResponse.CourseOption> getAvailablePrerequisiteOptions(Long courseId);
    
    boolean validatePrerequisiteAddition(Long courseId, Long prerequisiteCourseId);
}

File: frontend/src/services/prerequisiteCourseService.js
import API from './api.js';

export const prerequisiteCourseService = {
    // Enhanced endpoints for prerequisite course management
    getAllCoursesWithPrerequisiteInfo: async () => {
        return await API.get('/v1/prerequisite-course/courses/summary');
    },

    getCoursePrerequisiteDetails: async (courseId) => {
        return await API.get(`/v1/prerequisite-course/courses/${courseId}/details`);
    },

    getAvailablePrerequisiteOptions: async (courseId) => {
        return await API.get(`/v1/prerequisite-course/courses/${courseId}/available-prerequisites`);
    },

    updateCoursePrerequisites: async (courseId, updateData) => {
        return await API.put(`/v1/prerequisite-course/courses/${courseId}/prerequisites`, updateData);
    },

    validatePrerequisiteAddition: async (courseId, prerequisiteCourseId) => {
        return await API.get(`/v1/prerequisite-course/validate/${courseId}/prerequisite/${prerequisiteCourseId}`);
    },

    // Original endpoints (kept for compatibility)
    findAll: async () => {
        return await API.get('/v1/prerequisite-course/all');
    },

    findById: async (id) => {
        return await API.get(`/v1/prerequisite-course/details/${id}`);
    },

    save: async (prerequisiteCourseData) => {
        return await API.post('/v1/prerequisite-course/save', prerequisiteCourseData);
    },

    deleteById: async (id) => {
        return await API.delete(`/v1/prerequisite-course/delete/${id}`);
    }
}; 